src/Connection.php

```diff
@@ -106,125 +152,98 @@ public static function normalizeDSN($dsn, $user = null, $pass = null)
-                $c = new Connection_Oracle12(array_merge([
-                    'connection' => static::getPDO($dsn),
-                    'driverType' => $dsn['driverType'],
-                ], $args));
-
-                break;
-            case 'pgsql':
-                $c = new Connection_PgSQL(array_merge([
-                    'connection' => static::getPDO($dsn),
-                    'driverType' => $dsn['driverType'],
-                ], $args));
-
-                break;
-            case 'dumper':
-                $c = new Connection_Dumper(array_merge([
-                    'connection' => static::connect($dsn['rest'], $dsn['user'], $dsn['pass']),
-                ], $args));
-
-                break;
-            case 'counter':
-                $c = new Connection_Counter(array_merge([
-                    'connection' => static::connect($dsn['rest'], $dsn['user'], $dsn['pass']),
-                ], $args));
-
-                break;
-            // let PDO handle the rest
-            default:
-                $c = new static(array_merge([
-                    'connection' => static::connect(static::getPDO($dsn)),
-                ], $args));
+        /**
+         * @var Connection $connectionClass
+         */
+        $connectionClass = self::resolve($dsn['driverType']);
+
+        return new $connectionClass(array_merge([
+            'handler' => $connectionClass::createHandler($dsn),
+        ], $args));
+    }
+
+    /**
+     * Adds connection class to the registry for resolving in Connection::resolve method.
+     *
+     * Can be used as:
+     *
+     * Connection::register('mysql', MySQL\Connection::class), or
+     * MySQL\Connection::register()
+     *
+     * CustomDriver\Connection must be descendant of Connection class.
+     *
+     * @param string $driverType
+     * @param string $connectionClass
+     */
+    public static function register($driverType = null, $connectionClass = null)
+    {
+        if (!$connectionClass && is_a($driverType, self::class, true)) {
+            $connectionClass = $driverType;
+            $driverType = null;
+        }
+
+        $connectionClass = $connectionClass ?? static::class;
+
+        $driverType = $driverType ?? $connectionClass::DEFAULT_DRIVER_TYPE;
+
+        if (is_array($driverTypes = $driverType)) {
+            foreach ($driverTypes as $driverType => $connectionClass) {
+                if (is_numeric($driverType)) {
+                    $driverType = $connectionClass::DEFAULT_DRIVER_TYPE;
+                }
+
+                static::register($driverType, $connectionClass);
+            }
         }
 
-        return $c;
+        self::$registry[$driverType] = $connectionClass;
+    }
+
+    /**
+     * Resolves the connection class to use based on driver type.
+     *
+     * @param string $driverType
+     *
+     * @return string
+     */
+    public static function resolve($driverType)
```

Comment:
Too generic name

Reply:
Renamed to `resolveDriver`

Reply:
👍 wdyt about the same for registry, register() method..?

Reply:
I believe keeping only `register` is more clear and understandable.
`CustomConnection::register();` or
`CustomConnection::register('forSomeDriverType');` or 
`Connection::register('someDriverType', CustomConnection::class)`

Reply:
I meant `registerDriver`, dyt it is less clear?

Reply:
Yes, I understood about `registerDriver`. I think now is more clear but let me think a bit more about it.

Reply:
Or even `registerDriverClass`.

Reply:
Now is reformatted and more understandable I believe